### ISSUE TITLE
[Bug]: Copied/exported PNGs are blurry

### ISSUE THREAD

**Issue by ravindUwU** [tldraw/tldraw#5139]
### What happened?

I've noticed that when PNGs pasted into tldraw are exported/copied as PNG, the resultant image is blurry.

### How can we reproduce the bug?

In the project at https://www.tldraw.com/ro/nb7U7UwQt2AWvaoSoGecS?d=v108.-55.1912.1002.page I pasted a PNG image → right clicked → copy as → PNG and pasted it back in.

Please visit the link and zoom in to notice how the text is sharp on the left, and blurry on the right.
![](https://github.com/user-attachments/assets/b51d3bab-589d-49eb-b672-74b7fbc39522)

### What browsers are you seeing the problem on?

Firefox, Microsoft Edge

### Where did this happen?

The tldraw.com website

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by ravindUwU** [comment]
When a _single_ image shape containing a PNG is being copied/exported off tldraw, could the underlying image data be used as the clipboard/saved contents instead of the result of the canvas-based export, either via the existing "copy as → PNG" option, or a new "copy original"/"copy as → original" option 🤔?

This way the copied/exported image also wouldn't be affected by the padding added by the export (which is troublesome at times, and can't be switched off via the interface AFAIK). #3897 is about this and adds a programmatic way to remove the padding, but stops short of a UI for it (which I think would make a lot of sense).

**Comment by mimecuvalo** [comment]
This might be related to https://github.com/tldraw/tldraw/pull/5176
(or, @SomeHats I know you were looking at some retina/export-stuff maybe lately?)

**Comment by ravindUwU** [comment]
Does tldraw scale (?) contents of the canvas during the export? Could it be something to do with some sort of pixel ratio? I'm unsure what tldraw does here, so I'll refer to it as "scaling" 😅.

For example,

1. **The original picture (screenshot from this GitHub issue).**
   ![](https://github.com/user-attachments/assets/9852a2d8-8cb6-483e-a4a5-31c59f69e599)

2. **Same screenshot pasted into tldraw, then copied using Copy As > PNG.** The screenshot is blurry because it scaled.
   ![](https://github.com/user-attachments/assets/a9a331fc-28e8-4e59-b44e-c618ef23cd35)

3. **Same as 2, but annotated with text, ink, arrows.** tldraw annotations rendered nicely, but the screenshot is blurry. Scaling works ok with vectors like tldraw text, ink & arrows; but doesn't work with pictures pasted in.
   ![](https://github.com/user-attachments/assets/b0f95bee-c286-4522-8fe0-02434968d393)

**Comment by ravindUwU** [comment]
<img src="https://github.com/user-attachments/assets/9b47fecb-d6d1-4ac4-b4d6-3e9be10d0370" width="250">

---

Smol observation 👀, unsure if this has export implications: `getImageSize` assumes 72 DPI which might be different to that of the pasted image.

https://github.com/tldraw/tldraw/blob/24319e4fb9244a9b348c8a295b79698c39c7368a/packages/utils/src/lib/media/media.ts#L192-L201

https://stackblitz.com/edit/ravinduwu-gh-tldraw-5139-0?file=script.js has `600x400.72dpi.png` from <https://placehold.co/600x400.png> which is at 72 DPI, and `600x400.96dpi.png` which is a screenshot of the same size from my laptop at 96 DPI.

![](https://github.com/user-attachments/assets/e129cddb-ca2c-4c1a-ac43-52ebb33bb50c)

---

Pasting the 600x400 72 DPI image into tldraw and copying it as a PNG produces a 1200x800 image, on Firefox 143.0.3 and Edge 141.0.3537.57.. could this be something to do with the `pixelRatio = 2` defaults at 👇? Should this be 1, instead? 🤔

https://github.com/tldraw/tldraw/blob/24319e4fb9244a9b348c8a295b79698c39c7368a/packages/editor/src/lib/editor/Editor.ts#L9374

https://github.com/tldraw/tldraw/blob/24319e4fb9244a9b348c8a295b79698c39c7368a/packages/editor/src/lib/exports/getSvgAsImage.ts#L17

**Comment by steveruizok** [comment]
Can anyone reproduce this anymore?

**Comment by ravindUwU** [comment]
Yes, on Windows 11 running Firefox 150.0.2 and Edge 148.0.3967.54. The 72 DPI https://placehold.co/600x400.png pasted into tldraw and copied as a PNG produces a 1200x800 output (scaled 2x, and therefore, blurry).

![](https://github.com/user-attachments/assets/07ee7e53-670e-481f-b470-077c6891878e)

**Comment by ravindUwU** [comment]
I tried out the [preview](https://examples-git-fix-blurry-png-export.tldraw.club/) of #7937 which fixes the scaling issue.

![](https://github.com/user-attachments/assets/d32ebc29-cf0e-4774-8644-7486538e232f)

Seeing how [it doesn't look good on retina](https://github.com/tldraw/tldraw/pull/7937#issuecomment-3920113621), thoughts on enabling 2x only if a retina screen is detected (if that's even possible), exporting in 1x otherwise?

**Comment by mimecuvalo** [comment]
@kostyafarber this was the temporary Chrome issue I believe? can you confirm and close if so?

**Comment by kostyafarber** [comment]
don't think this is related, this is about scaling the logical pixels, not the image dimensions itself, doubling the pixels on non retina screens will cause blurriness but at the expense retina screens that actually have more pixels, we could experiment with trying to detect DPR from the browser and changing that number based on it
